Core/Src/pdu.c

```diff
@@ -15,26 +16,75 @@
 
 #define SHUTDOWN_ADDR PCA_I2C_ADDR_3
 #define CTRL_ADDR     PCA_I2C_ADDR_2
-#define RTDS_DURATION 2500
+#define RTDS_DURATION 2500 /* ms at 1kHz tick rate */
 
 static osMutexAttr_t pdu_mutex_attributes;
 
-static void rtds_shutoff_cb(void *pv_params)
+static uint8_t sound_rtds(pdu_t *pdu)
+{
+	if (!pdu)
+		return -1;
+
+	osStatus_t stat = osMutexAcquire(pdu->mutex, MUTEX_TIMEOUT);
+	if (stat)
+		return stat;
+
+	/* write RTDS over i2c */
+	HAL_StatusTypeDef error = pca9539_write_pin(
+		pdu->ctrl_expander, PCA_OUTPUT_1_REG, RTDS_CTRL, true);
+	if (error != HAL_OK) {
+		osMutexRelease(pdu->mutex);
+		return error;
+	}
+
+	osMutexRelease(pdu->mutex);
+
+	return 0;
+}
+
+static uint8_t rtds_shutoff(void *pv_params)
 {
 	pdu_t *pdu = (pdu_t *)pv_params;
 	osStatus_t stat = osMutexAcquire(pdu->mutex, MUTEX_TIMEOUT);
 	if (stat)
-		return;
+		return stat;
 
 	/* write RTDS over i2c */
 	HAL_StatusTypeDef error = pca9539_write_pin(
 		pdu->ctrl_expander, PCA_OUTPUT_1_REG, RTDS_CTRL, false);
 	if (error != HAL_OK) {
 		osMutexRelease(pdu->mutex);
-		return;
+		return error;
 	}
 
 	osMutexRelease(pdu->mutex);
+	return 0;
+}
+
+osThreadId_t rtds_thread;
+const osThreadAttr_t rtds_attributes = { .name = "RtdsThread",
+					 .stack_size = 300,
+					 /* The task will run infrequently */
+					 .priority = osPriorityRealtime7 };
+void vRTDS(void *arg)
+{
+	pdu_t *pdu = (pdu_t *)arg;
+
+	fault_data_t rtds_fault = { .id = RTDS_FAULT, .severity = DEFCON4 };
+
+	for (;;) {
+		osThreadFlagsWait(SOUND_RTDS_FLAG, osFlagsWaitAny,
+				  osWaitForever);
+		if (sound_rtds(pdu)) {
+			rtds_fault.diag = "Unable to sound RTDS";
+			queue_fault(&rtds_fault);
+		}
+		osDelay(RTDS_DURATION);
```

Comment:
Im very hesistant of a rules referenced time guarrantee to be an OS delay.  I feel like a timer would be more resilient but feel free to correct me if im wrong.

Reply:
I had to make the RTDS task since the RTDS functions use mutexes so we can't use a timer for them. Also the delay is not an issue because the task has the highest possible RTOS priority, so it will run right when the delay is over. I will time it to make sure it is rules compliant.

Reply:
We can use a mutex if the timeout is 0.